### PR TITLE
feat(app-cmds): Implement User-Installed Apps

### PR DESCRIPTION
**Pull request by Kevin-OVI**
## Summary

This PR implements the "integration_types" and "contexts" fields on app commands to support User-Installed Apps. It also deprecates the "dm_permission" parameter in all application commands, as it has been replaced by "contexts".
The PR is marked as draft because user apps are still in preview.

## This is a **Code Change**

- [X] I have tested my changes.
- [X] I have updated the documentation to reflect the changes.
- [X] I have run `task pyright` and fixed the relevant issues.

### CODE REVIEW

File: nextcord/enums.py
Comment (spifory) [inline nextcord/enums.py:2042]: Both class docstrings (and their parameters' docstrings) added in this file are missing periods at the ends.

**Review by spifory — CHANGES_REQUESTED** [review]
You likely want to take a look at https://github.com/discord/discord-api-docs/pull/6723 to implement as it seems a lot is missing in this PR (I do understand this is still a draft though)

**Comment by Kevin-OVI** [comment]
> You likely want to take a look at [discord/discord-api-docs#6723](https://github.com/discord/discord-api-docs/pull/6723) to implement as it seems a lot is missing in this PR (I do understand this is still a draft though)

I only had implemented what I had seen as necessary to enable user installed apps, without checking what had actually changed in the change-log. I think everything from there is now implemented.

**Comment by Kevin-OVI** [comment]
I see that the pre-commit.ci check is failing. Is there anyway I can fix the rust compiler not being installed ? I ran `task lint` on my side before committing so there shouldn't be any issues.

**Comment by alentoghostflame** [comment]
> I see that the pre-commit.ci check is failing. Is there anyway I can fix the rust compiler not being installed ? I ran `task lint` on my side before committing so there shouldn't be any issues.

I don't believe so, we have to wait for a `libcst` wheel to be published for whatever platform + version the pre-commit.ci check uses.

**Review by EmreTech — CHANGES_REQUESTED** [review]
Looks good otherwise!

**Review by spifory — CHANGES_REQUESTED** [review]
I seem to have made this review a month ago and forgot, but anyways just some nitpicks in the code

File: nextcord/application_command.py
Comment (teaishealthy) [inline nextcord/application_command.py:2011]:
```suggestion
            Where the command can be used, only for globally-scoped commands. By default, all interaction context types are included for new commands.
```

File: nextcord/application_command.py
Comment (teaishealthy) [inline nextcord/application_command.py:2039]: DeprecationWarning?

**Review by teaishealthy — CHANGES_REQUESTED** [review]

File: nextcord/message.py
Comment (EmreTech) [inline nextcord/message.py:739]:
```suggestion
        return f"<{self.__class__.__name__} id={self.id} type={self.type} user={self.user!r} name={self.name!r}>"
```

**Review by EmreTech — COMMENTED** [review]
Tested! 🎉

File: nextcord/application_command.py
Comment (Kevin-OVI) [reply nextcord/application_command.py:2039]: I was thinking this argument could be removed as it's a breaking change and the next version would be 3.0 anyway.

**Comment by EmreTech** [comment]
This may need updates from the new [changelog](https://discord.com/developers/docs/change-log#june-27-2024-api-updates).

**Comment by tizu69** [comment]
I'm not too familiar with this, is thre some way to use this already?

**Comment by xXenvy** [comment]
> I'm not too familiar with this, is thre some way to use this already?

If you want to play with it, go ahead, but if you want to use it for ‘production’ - I would strongly suggest that you don't.

To install: `pip install -U git+https://github.com/Kevin-OVI/nextcord.git@user-apps`
U may also need docs: https://nextcord--1181.org.readthedocs.build/en/1181/

**Comment by alentoghostflame** [comment]
@Kevin-OVI, are you still working on this? Once you get the conflicts resolved, I'll take a serious look at getting this merged in.

**Review by alentoghostflame — APPROVED** [review]
`dm_permission` either needs to be removed or a `DeprectionWarning` added (IMO it should be removed) but overall this PR looks really really good! While I haven't scrutinized every bit of the code, it seems like all newly added bits are in the correct files. Application command payloads compare correctly, and (after merging #1204) even lazy loading is working correctly!

Overall, this works nicely and everything seems to function as expected. A decision with `dm_permission` needs to be made and I would like to have at least one more solid approval before merging this in.

File: nextcord/application_command.py
Comment (Kevin-OVI) [reply nextcord/application_command.py]: I've removed the `dm_permission` parameter. The PR should be ready to merge.

**Review by EmmmaTech — APPROVED** [review]
Looks good :3

**Comment by alentoghostflame** [comment]
Alright, if anything goes badly @EmmmaTech is to blame obv.

**Comment by EmmmaTech** [comment]
> Alright, if anything goes badly @EmmmaTech is to blame obv.

😭

**Pull request merged**
